internal/health/health.go

```diff
@@ -146,24 +148,29 @@ func (h *Health) Ready(source string, ready bool) {
 	}
 	h.readies[source] = ready
 	h.timeLeft[source] = h.timeouts[source]
+	if !h.alives[source] {
+		h.alives[source] = true
+		h.Logger.Info().WithField("source", source).Logf("Health.Ready reporting source alive")
+	}
 	h.Metrics.Gauge("is_ready", h.checkReady())
 	h.Metrics.Gauge("is_alive", h.checkAlive())
 }
 
 // IsAlive returns true if all registered services are alive
 func (h *Health) IsAlive() bool {
-	h.mut.RLock()
-	defer h.mut.RUnlock()
+	h.mut.Lock()
+	defer h.mut.Unlock()
 	return h.checkAlive()
 }
 
 // checkAlive returns true if all registered services are alive
-// only call with the lock held
+// only call with a write lock held
 func (h *Health) checkAlive() bool {
 	// if any counter is 0, we're dead
 	for source, a := range h.timeLeft {
-		if a == 0 {
+		if a == 0 && h.alives[source] {
```

Comment:
It looks like we will only return false the first time we transition from alive to not alive.  If we are non alive twice in a row we wont return false.

Reply:
Do you mean if `checkAlive` is called twice and `Ready` is not called at all?

Reply:
If `checkAlive` is called twice without `h.alives[source]` being set back to `true`.

Reply:
Thanks. Fixed.